common/src/int.rs

```diff
@@ -1,120 +1,631 @@
+use std::{
+    cmp::Ordering,
+    ops::{
+        Add, AddAssign, BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXor, BitXorAssign, Div,
+        DivAssign, Mul, MulAssign, Neg, Not, Rem, RemAssign, Sub, SubAssign,
+    },
+};
+
 use bstr::ByteSlice;
-use num_bigint::{BigInt, BigUint, Sign};
-use num_traits::{ToPrimitive, Zero};
+use derive_more::{
+    Binary, Display, From, FromStr, LowerHex, Octal, Product, Shl, ShlAssign, Shr, ShrAssign, Sum,
+    UpperHex,
+};
+use malachite::{
+    num::{
+        arithmetic::traits::{
+            Abs, DivMod, DivRem, DivisibleBy, ExtendedGcd, FloorSqrt, Gcd, Lcm, Mod, ModPow,
+            Parity, Sign,
+        },
+        conversion::traits::{
+            FromStringBase, OverflowingInto, PowerOf2Digits, RoundingInto, ToStringBase,
+        },
+        logic::traits::{CountOnes, SignificantBits},
+    },
+    rounding_modes::RoundingMode,
+    Integer, Natural, Rational,
+};
+use num_traits::{Num, One, Pow, Signed, ToPrimitive, Zero};
 
-pub fn bytes_to_int(lit: &[u8], mut base: u32) -> Option<BigInt> {
-    // split sign
-    let mut lit = lit.trim();
-    let sign = match lit.first()? {
-        b'+' => Some(Sign::Plus),
-        b'-' => Some(Sign::Minus),
-        _ => None,
+#[repr(transparent)]
+#[derive(
+    Hash,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Clone,
+    FromStr,
+    Display,
+    Binary,
+    Octal,
+    LowerHex,
+    UpperHex,
+    From,
+    Shr,
+    Shl,
+    Sum,
+    Product,
+    ShlAssign,
+    ShrAssign,
+)]
+#[from(forward)]
+pub struct BigInt(Integer);
```

Comment:
It seems you are almost making alternative num-bigint implementation. 
I think making a crate like `malachite-bigint` as a drop-in replacement of `num-bigint` might be worth if you are interested in.

Reply:
malachite-bigint is a good idea